elasticapm/utils/stacks.py

```diff
@@ -163,23 +163,31 @@ def to_dict(dictish):
     return dict((k, dictish[k]) for k in m())
 
 
-def iter_traceback_frames(tb):
+def iter_traceback_frames(tb, config=None):
     """
     Given a traceback object, it will iterate over all
     frames that do not contain the ``__traceback_hide__``
     local variable.
     """
+    max_frames = config.stack_trace_limit if config else -1
+    if not max_frames:
+        return
+    frames = []
     while tb:
         # support for __traceback_hide__ which is used by a few libraries
         # to hide internal frames.
         frame = tb.tb_frame
-        f_locals = getattr(frame, "f_locals", {})
-        if not _getitem_from_frame(f_locals, "__traceback_hide__"):
-            yield frame, getattr(tb, "tb_lineno", None)
+        f_locals = getattr(frame, "f_locals", None)
+        if f_locals is None or not _getitem_from_frame(f_locals, "__traceback_hide__"):
+            frames.append((frame, getattr(tb, "tb_lineno", None)))
         tb = tb.tb_next
+    if max_frames != -1:
+        frames = frames[-max_frames:]
```

Comment:
Is it worth counting the number of frames in `tb` first, so you can skip and then yield as before? Then you wouldn't need to create a list only to discard some of the entries.

Reply:
Sorry for the late answer. I'd rather not traverse the linked list of frames twice (once for counting, once for yielding, unless I'm missing something), as attribute lookup in Python can be expensive. Creating the list should be OK, as most tracebacks don't have an unreasonable amount of frames.